MoltenVK/MoltenVK/Commands/MVKCommandEncoderState.mm

```diff
@@ -337,130 +266,255 @@
-                                                     bool isDynamic) {
-    // Abort if we are using dynamic, but call is not dynamic.
-	if ( !isDynamic && _cmdEncoder->supportsDynamicState(VK_DYNAMIC_STATE_BLEND_CONSTANTS) ) { return; }
-
-    if (_red != red || _green != green || _blue != blue || _alpha != alpha) {
-        markDirty();
-        _red = red;
-        _green = green;
-        _blue = blue;
-        _alpha = alpha;
-    }
+	MVKMTLScissors mtlScissors = isDynamic ? _mtlScissors[StateScope::Dynamic] : _mtlScissors[StateScope::Static];
+	size_t sCnt = min((uint32_t)scissors.size(), maxScissors - firstScissor);
+	for (uint32_t sIdx = 0; sIdx < sCnt; sIdx++) {
+		mtlScissors.scissors[firstScissor + sIdx] = mvkMTLScissorRectFromVkRect2D(scissors[sIdx]);
+		mtlScissors.scissorCount = max(mtlScissors.scissorCount, sIdx + 1);
+	}
+	setContent(Scissors);
 }
 
-void MVKBlendColorCommandEncoderState::encodeImpl(uint32_t stage) {
-    if (stage != kMVKGraphicsStageRasterization) { return; }
-    [_cmdEncoder->_mtlRenderEncoder setBlendColorRed: _red green: _green blue: _blue alpha: _alpha];
+void MVKRenderingCommandEncoderState::setRasterizerDiscardEnable(VkBool32 rasterizerDiscardEnable, bool isDynamic) {
+	bool mtlRasterizerDiscardEnable = static_cast<bool>(rasterizerDiscardEnable);
+	setContent(RasterizerDiscardEnable);
+}
+
+#pragma mark Encoding
+
+void MVKRenderingCommandEncoderState::encodeImpl(uint32_t stage) {
+	if (stage != kMVKGraphicsStageRasterization) { return; }
+
+	auto& rendEnc = _cmdEncoder->_mtlRenderEncoder;
+
+	if (isDirty(CullMode)) { [rendEnc setCullMode: getContent(CullMode)]; }
+	if (isDirty(FrontFace)) { [rendEnc setFrontFacingWinding: getContent(FrontFace)]; }
+	if (isDirty(BlendConstants)) {
+		auto& bcFlt = getContent(BlendConstants).float32;
+		[rendEnc setBlendColorRed: bcFlt[0] green: bcFlt[1] blue: bcFlt[2] alpha: bcFlt[3]];
+	}
+	if (isDirty(DepthBiasEnable) || isDirty(DepthBias)) {
+		if (getContent(DepthBiasEnable)) {
+			auto& db = getContent(DepthBias);
+			[rendEnc setDepthBias: db.depthBiasConstantFactor
+					   slopeScale: db.depthBiasSlopeFactor
+							clamp: db.depthBiasClamp];
+		} else {
+			[rendEnc setDepthBias: 0 slopeScale: 0 clamp: 0];
+		}
+	}
+	if (isDirty(DepthClipEnable) && getDevice()->_enabledFeatures.depthClamp) {
+		[rendEnc setDepthClipMode: getContent(DepthClipEnable)];
+	}
+
+	if (isDirty(StencilReference)) {
+		auto& sr = getContent(StencilReference);
+		[rendEnc setStencilFrontReferenceValue: sr.frontFaceValue backReferenceValue: sr.backFaceValue];
+	}
+
+	if (isDirty(Viewports)) {
+		auto& mtlViewports = getContent(Viewports);
+		if (_cmdEncoder->_pDeviceFeatures->multiViewport) {
+#if MVK_MACOS_OR_IOS
+			[rendEnc setViewports: mtlViewports.viewports count: mtlViewports.viewportCount];
+#endif
+		} else {
+			[rendEnc setViewport: mtlViewports.viewports[0]];
+		}
+	}
+
+	// If rasterizing discard has been dynamically enabled, or culling has been dynamically 
```

Comment:
Clever hack.

Reply:
Thanks. I was not at all sure it would work. But it got me through CTS successfully.

I am hopeful that Apple Silicon can optimize to apply the scissor test prior shader processing (without needing to force it with `early_fragment_tests`), which would help avoid any side-effects of shader execution (including performance degradation). It seems likely that it would optimize ownership and scissor tests that way.